ADR: Align on Zod version to safely support exposing Zod schemas from plugins (#22089)
backstage/backstage: docs/architecture-decisions/adr014-exposing-zod-schemas.md

---
id: adrs-adr014
title: 'ADR014: Exposing Zod schemas'
# prettier-ignore
description: Architecture Decision Record (ADR) for treating Zod as a first class dependency and support for exposing Zod schemas from plugins.
---

## Context

Backstage currently uses the [Zod](https://zod.dev) library extensively for creating schemas to validate data types and use these schemas as TypeScript types within a plugin (and sometimes as parameters to public APIs). However, we are currently not exposing these schemas directly as part of plugin public APIs.

> Plugins that are working with the data types of other plugin, can benefit from these schemas for performing their own validation and composing more complex types. As an example, the permission framework has the concept of a `Permission` data type, which internally is validated with a Zod schema.

Reviewer: ```suggestion
Plugins that are working with the data types of other plugins, can benefit from these schemas for performing their own validation and composing more complex types. As an example, the permission framework has the concept of a `Permission` data type, which internally is validated with a Zod schema.
```

```ts
const permissionSchema = z.union([
  z.object({
    type: z.literal('basic'),
    name: z.string(),
    attributes: z
      .union([
        z.literal('create'),
        z.literal('read'),
        z.literal('update'),
        z.literal('delete'),
      ])
      .optional(),
  }),
  z.object({
    type: z.literal('resource'),
    name: z.string(),
    attributes: attributesSchema,
    resourceType: z.string(),
  }),
]);
```

> Adopters of Backstage may want to create an RBAC frontend interface for the permission framework, which would require to work with the `Permission` data type. They would benefit from aligning with the validation function from the permission plugin, and be able to compose additional schemas from it. For example.

Reviewer: ```suggestion
Adopters of Backstage may want to create an RBAC frontend interface for the permission framework, which would require to work with the `Permission` data type. They would benefit from being able to use the validation function from the permission plugin, and being able to compose additional schemas from it. For example.
```

```ts
import { permissionSchema } from '@backstage/permission-node';

const rbacRolePermissions = z.union({
  permissions: z.array(permissionSchema),
});
```

However, there is a possible risk to exposing the Zod schemas. A difference in Zod version could result in creating a schema that doesn't align with the expected data. It also externally couples Backstage to the Zod library.

## Decision

> 1. Zod is the preferred library for validating data, and that plugins should where possible align with using Zod.

Author: After the discussions, seeing the schemas and the issues that are brought up in [this PR](https://github.com/backstage/backstage/pull/22205) and how we have to use tooling to generate things, as opposed to how simple to write, use and composes schemas with Zod are (shown in [this PR](https://github.com/backstage/backstage/pull/22080)), I don't understand the value in OpenAPI other than it's a technology agnostic spec, and that we're sacrificing developer experience for it. 

However I'm happy to update the ADR to say OpenAPI is a preference if it gets us closer to being able to export zod schemas.

> 1. Zod is the preferred library for validating data, and that plugins should where possible align with using Zod.

Reviewer 2: The value of OpenAPI is in the secondary effects, like data validation, client generation, documentation and testing. Looking just at writing the schema is missing the key value proposition. These schemas are rarely changed and there's nothing stopping you from translating the schema to zod to then use. My concern here was that if we push zod as the preferred data validation library it would steer folks away from the secondary effects and they would end up in the same problems of hand maintained client, duplicated validation and types, and unstable type safety between client and server that I'm addressing with tooling around OpenAPI.

It might be helpful with an example of where the line would be. Looking through the meeting notes, I'm still not sure where that line would end up being/what would sit in OpenAPI and what would be exported with Zod.

The `catalog-model` entity definitions could be a good starting place where the definition is no longer definable by JSON schema as it goes through processors and whatnot. The OpenAPI definition for that is just `additionalProperties: true` and we surface errors from the processing loop. I _think_ would make sense to update to zod and export as zod definitions instead of JSON schema.

2. Adopters of Backstage should keep their version of Zod in sync within the same `major` version of Zod used inside Backstage. This is to ensure data types and validation remain compatible.

> 3. The second step above opens up the possibility to safely expose Zod schemas from plugins.

Reviewer 3: I think this can just be a statement that we will start directly exposing Zod schemas in our APIs
